src/lexer.coffee

```diff
@@ -141,34 +141,38 @@ exports.Lexer = class Lexer
       else
         'IDENTIFIER'
 
-    if tag is 'IDENTIFIER' and (id in JS_KEYWORDS or id in COFFEE_KEYWORDS) and
-       not (@exportSpecifierList and id in COFFEE_KEYWORDS)
-      tag = id.toUpperCase()
-      if tag is 'WHEN' and @tag() in LINE_BREAK
-        tag = 'LEADING_WHEN'
-      else if tag is 'FOR'
-        @seenFor = yes
-      else if tag is 'UNLESS'
-        tag = 'IF'
-      else if tag is 'IMPORT'
-        @seenImport = yes
-      else if tag is 'EXPORT'
-        @seenExport = yes
-      else if tag in UNARY
-        tag = 'UNARY'
-      else if tag in RELATION
-        if tag isnt 'INSTANCEOF' and @seenFor
-          tag = 'FOR' + tag
-          @seenFor = no
-        else
-          tag = 'RELATION'
-          if @value() is '!'
-            poppedToken = @tokens.pop()
-            id = '!' + id
-
-    if tag is 'IDENTIFIER' and id in RESERVED
-      @error "reserved word '#{id}'", length: id.length
-
+    if tag is 'IDENTIFIER'
+      if @seenFor and id is 'from'
```

Comment:
Not sure. You can copy the `tag is 'IDENTIFIER'` into each of the ifs, but then the `else if` condition becomes comically long (see below).

I do agree that probably the if at the bottom of the big if (`if id in RESERVED`) should broken out and returned to the main body of the method.

```coffee
if tag is 'IDENTIFIER' and @seenFor and id is 'from'
  tag = 'FORFROM'
  @seenFor = no
else if tag is 'IDENTIFIER' and (id in JS_KEYWORDS or id in COFFEE_KEYWORDS) and
   not (@exportSpecifierList and id in COFFEE_KEYWORDS)
  tag = id.toUpperCase()
  if tag is 'WHEN' and @tag() in LINE_BREAK
    tag = 'LEADING_WHEN'
  else if tag is 'FOR'
    @seenFor = yes
  else if tag is 'UNLESS'
    tag = 'IF'
  else if tag is 'IMPORT'
    @seenImport = yes
  else if tag is 'EXPORT'
    @seenExport = yes
  else if tag in UNARY
    tag = 'UNARY'
  else if tag in RELATION
    if tag isnt 'INSTANCEOF' and @seenFor
      tag = 'FOR' + tag
      @seenFor = no
    else
      tag = 'RELATION'
      if @value() is '!'
        poppedToken = @tokens.pop()
        id = '!' + id

if tag is 'IDENTIFIER' and id in RESERVED
  @error "reserved word '#{id}'", length: id.length
```